ssl/quic/quic_wire.c

```diff
@@ -0,0 +1,776 @@
+{
+    if (!expect_frame_header(pkt, OSSL_QUIC_FRAME_TYPE_CRYPTO)
+            || !PACKET_get_quic_vlint(pkt, &f->offset)
+            || !PACKET_get_quic_vlint(pkt, &f->len))
+        return 0;
+
+    if (PACKET_remaining(pkt) < f->len)
+        return 0;
+
+    f->data = PACKET_data(pkt);
+
+    if (!PACKET_forward(pkt, f->len))
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_decode_frame_new_token(PACKET         *pkt,
+                                          const uint8_t **token,
+                                          size_t         *token_len)
+{
+    uint64_t token_len_;
+
+    if (!expect_frame_header(pkt, OSSL_QUIC_FRAME_TYPE_NEW_TOKEN)
+            || !PACKET_get_quic_vlint(pkt, &token_len_))
+        return 0;
+
+    if (token_len_ > SIZE_MAX)
+        return 0;
+
+    *token      = PACKET_data(pkt);
+    *token_len  = token_len_;
+
+    if (!PACKET_forward(pkt, token_len_))
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_decode_frame_stream(PACKET *pkt,
+                                       OSSL_QUIC_FRAME_STREAM *f)
+{
+    uint64_t frame_type;
+
+    if (!expect_frame_header_mask(pkt, OSSL_QUIC_FRAME_TYPE_STREAM,
+                                  OSSL_QUIC_FRAME_FLAG_STREAM_MASK,
+                                  &frame_type)
+            || !PACKET_get_quic_vlint(pkt, &f->stream_id))
+        return 0;
+
+    if ((frame_type & OSSL_QUIC_FRAME_FLAG_STREAM_OFF) != 0) {
+        if (!PACKET_get_quic_vlint(pkt, &f->offset))
+            return 0;
+    } else {
+        f->offset = 0;
+    }
+
+    f->has_len = ((frame_type & OSSL_QUIC_FRAME_FLAG_STREAM_LEN) != 0);
+    f->is_fin  = ((frame_type & OSSL_QUIC_FRAME_FLAG_STREAM_FIN) != 0);
+
+    if (f->has_len) {
+        if (!PACKET_get_quic_vlint(pkt, &f->len))
+            return 0;
+
+        if (f->len > PACKET_remaining(pkt))
+            return 0;
+    } else {
+        f->len = PACKET_remaining(pkt);
+    }
+
+    f->data = PACKET_data(pkt);
+
+    if (!PACKET_forward(pkt, f->len))
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_decode_frame_max_data(PACKET *pkt,
+                                         uint64_t *max_data)
+{
+    if (!expect_frame_header(pkt, OSSL_QUIC_FRAME_TYPE_MAX_DATA)
+            || !PACKET_get_quic_vlint(pkt, max_data))
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_decode_frame_max_stream_data(PACKET *pkt,
+                                                uint64_t *stream_id,
+                                                uint64_t *max_stream_data)
+{
+    if (!expect_frame_header(pkt, OSSL_QUIC_FRAME_TYPE_MAX_STREAM_DATA)
+            || !PACKET_get_quic_vlint(pkt, stream_id)
+            || !PACKET_get_quic_vlint(pkt, max_stream_data))
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_decode_frame_max_streams(PACKET *pkt,
+                                            uint64_t *max_streams)
+{
+    if (!expect_frame_header_mask(pkt, OSSL_QUIC_FRAME_TYPE_MAX_STREAMS_BIDI,
```

Comment:
This handles both. `expect_frame_header_mask` masks off the LSB, and therefore decodes both types.

There is no output like `int *is_uni` because in order for a user to have called `ossl_quic_wire_decode_max_frame_streams` in the first place, they must already know the frame type, and seeing whether the type is `..._BIDI` or `..._UNI` is trivial for them to do.

Reply:
Perhaps a comment explaining some of that?